src/psyclone/psyir/transformations/intrinsics/sum2code_trans.py

```diff
@@ -0,0 +1,361 @@
+    summed:
+
+    .. code-block:: python
+
+        R = SUM(ARRAY, dimension=2)
+
+    If the array is two dimensional, the equivalent code
+    for real data is:
+
+    .. code-block:: python
+
+        R(:) = 0.0
+        DO J=LBOUND(A,2),UBOUND(A,2)
+          DO I=LBOUND(A,1),UBOUND(A,1)
+            R(I) += R(I) + A(I,J)
+
+    If the mask argument is provided then the mask is used to
+    determine whether the sum is applied:
+
+    .. code-block:: python
+
+        R = SUM(ARRAY, mask=MOD(array, 2.0)==1)
+
+    If the array is two dimensional, the equivalent code
+    for real data is:
+
+    .. code-block:: python
+
+        R = 0.0
+        DO J=LBOUND(A,2),UBOUND(A,2)
+          DO I=LBOUND(A,1),UBOUND(A,1)
+            if (MOD(array(I,J), 2.0)==1):
+              R += R + A(I,J)
+
+    '''
+    def __init__(self):
+        super().__init__()
+        self._operator_name = "SUM"
+        self._classes = (UnaryOperation, BinaryOperation, NaryOperation)
+        self._operators = (UnaryOperation.Operator.SUM,
+                           BinaryOperation.Operator.SUM,
+                           NaryOperation.Operator.SUM)
+
+    def validate(self, node, options=None):
+        '''Check that the input node is valid before applying the
+        transformation.
+
+        :param node: a Sum Operation. This could be a UnaryOperation, \
+            BinaryOperation or NaryOperation depending on how many \
+            arguments it has.
+        :type node: :py:test:`psyclone.psyir.nodes.operation.UnaryOperation` \
+            | :py:test:`psyclone.psyir.nodes.operation.BinaryOperation` | \
+            :py:test:`psyclone.psyir.nodes.operation.NaryOperation`
+        :param options: options for the transformation.
+        :type options: Optional[Dict[str,str]]
+
+        :raises TransformationError: if a valid value for the \
+            dimension argument can't be determined.
+        :raises TransformationError: if the array argument is not an array.
+        :raises TransformationError: if the array argument is not an array.
+        :raises TransformationError: if the shape of the array is not \
+            supported.
+        :raises TransformationError: if the array datatype is not \
+            supported.
+
+        '''
+        # Avoid circular import error
+        # pylint: disable=import-outside-toplevel
+        from psyclone.psyir.transformations import TransformationError
+
+        super().validate(node, options)
+
+        array_ref, dim_ref, _ = get_args(node)
+        if dim_ref and not isinstance(dim_ref, (Literal, Reference)):
+            raise TransformationError(
+                f"Can't find the value of the dimension argument. Expected "
+                f"it to be a literal or a reference but found "
+                f"'{self._writer(dim_ref)}' which is a "
+                f"'{type(dim_ref).__name__}'.")
+
+        if len(array_ref.children) == 0:
```

Comment:
Please add tests with this 2 cases whatever the answers to the questions are.

Reply:
Well I did say it was a little rough around the edges and here is an edge :-). I've added validation checks so that appropriate exceptions are raised. The latter (literal dimension values) are actually invalid for the sum intrinsic.